File: 2024/07/31/105. Construct Binary Tree from Preorder and Inorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
            inorder_idx = {inorder[i]: i for i in range(len(inorder))}
            def build_tree(preorder, inorder_start, inorder_end):
                if not preorder or inorder_start < 0 or inorder_end > len(inorder):
                    return None

                root_val = preorder[0]
                root_inorder_idx = inorder_idx[root_val]
                if root_inorder_idx > inorder_end or root_inorder_idx < inorder_start:
                    return None
                root = TreeNode(preorder.pop(0))
                root.left = build_tree(preorder, inorder_start, root_inorder_idx - 1)
                root.right = build_tree(preorder, root_inorder_idx + 1, inorder_end)

                return root
            return build_tree(preorder, 0, len(inorder) - 1)
        

File: 2024/07/31/105. Construct Binary Tree from Preorder and Inorder Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorder_idx;
        for (int i = 0; i < inorder.size(); ++i) {
            inorder_idx[inorder[i]] = i;
        }
        
        return build_tree(preorder, inorder, 0, inorder.size() - 1, inorder_idx);
    }
    
private:
    TreeNode* build_tree(vector<int>& preorder, vector<int>& inorder, int inorder_start, int inorder_end, unordered_map<int, int>& inorder_idx) {
        if (preorder.empty() || inorder_start < 0 || inorder_end >= inorder.size() || inorder_start > inorder_end) {
            return nullptr;
        }
        
        int root_val = preorder[0];
        preorder.erase(preorder.begin());
        
        int root_inorder_idx = inorder_idx[root_val];
        if (root_inorder_idx > inorder_end || root_inorder_idx < inorder_start) {
            return nullptr;
        }
        
        TreeNode* root = new TreeNode(root_val);
        root->left = build_tree(preorder, inorder, inorder_start, root_inorder_idx - 1, inorder_idx);
        root->right = build_tree(preorder, inorder, root_inorder_idx + 1, inorder_end, inorder_idx);
        
        return root;
    }
};